src/data/gaussian_int.lean

```diff
@@ -0,0 +1,170 @@
+/-
+Copyright (c) 2019 Chris Hughes. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Chris Hughes
+
+The gaussian integers ℤ[i].
+-/
+
+import number_theory.pell data.complex.basic algebra.euclidean_domain algebra.associated
+open zsqrtd complex
+
+@[reducible] def gaussian_int : Type := zsqrtd (-1)
+
+local notation `ℤ[i]` := gaussian_int
+
+namespace gaussian_int
+
+instance : has_repr ℤ[i] := ⟨λ x, "⟨" ++ repr x.re ++ ", " ++ repr x.im ++ "⟩"⟩
+
+instance : comm_ring ℤ[i] := zsqrtd.comm_ring
+
+def norm (x : ℤ[i]) : ℕ := x.re.nat_abs * x.re.nat_abs + x.im.nat_abs * x.im.nat_abs
+
+def to_complex (x : ℤ[i]) : ℂ := x.re + x.im * I
+
+instance : has_coe (ℤ[i]) ℂ := ⟨to_complex⟩
+
+lemma to_complex_def (x : ℤ[i]) : (x : ℂ) = x.re + x.im * I := rfl
+
+lemma to_complex_def' (x y : ℤ) : ((⟨x, y⟩ : ℤ[i]) : ℂ) = x + y * I := by simp [to_complex_def]
+
+lemma to_complex_def₂ (x : ℤ[i]) : (x : ℂ) = ⟨x.re, x.im⟩ :=
+by apply complex.ext; simp [to_complex_def]
+
+instance to_complex.is_ring_hom : is_ring_hom to_complex:=
+by refine_struct {..}; intros; apply complex.ext; simp [to_complex]
+
+instance : is_ring_hom (coe : ℤ[i] → ℂ) := to_complex.is_ring_hom
+
+@[simp] lemma to_real_re (x : ℤ[i]) : ((x.re : ℤ) : ℝ) = (x : ℂ).re := by simp [to_complex_def]
+@[simp] lemma to_real_im (x : ℤ[i]) : ((x.im : ℤ) : ℝ) = (x : ℂ).im := by simp [to_complex_def]
+@[simp] lemma to_complex_re (x y : ℤ) : ((⟨x, y⟩ : ℤ[i]) : ℂ).re = x := by simp [to_complex_def]
+@[simp] lemma to_complex_im (x y : ℤ) : ((⟨x, y⟩ : ℤ[i]) : ℂ).im = y := by simp [to_complex_def]
+@[simp] lemma to_complex_add (x y : ℤ[i]) : ((x + y : ℤ[i]) : ℂ) = x + y := is_ring_hom.map_add coe
+@[simp] lemma to_complex_mul (x y : ℤ[i]) : ((x * y : ℤ[i]) : ℂ) = x * y := is_ring_hom.map_mul coe
+@[simp] lemma to_complex_one : ((1 : ℤ[i]) : ℂ) = 1 := is_ring_hom.map_one coe
+@[simp] lemma to_complex_zero : ((0 : ℤ[i]) : ℂ) = 0 := is_ring_hom.map_zero coe
+@[simp] lemma to_complex_neg (x : ℤ[i]) : ((-x : ℤ[i]) : ℂ) = -x := is_ring_hom.map_neg coe
+@[simp] lemma to_complex_sub (x y : ℤ[i]) : ((x - y : ℤ[i]) : ℂ) = x - y := is_ring_hom.map_sub coe
+
+@[simp] lemma to_complex_inj {x y : ℤ[i]} : (x : ℂ) = y ↔ x = y :=
+by cases x; cases y; simp [to_complex_def₂]
+
+@[simp] lemma to_complex_eq_zero {x : ℤ[i]} : (x : ℂ) = 0 ↔ x = 0 :=
+by rw [← to_complex_zero, to_complex_inj]
+
+@[simp] lemma nat_cast_real_norm (x : ℤ[i]) : (x.norm : ℝ) = (x : ℂ).norm_sq :=
+by rw [norm, nat.cast_add, ← int.cast_coe_nat, ← int.cast_coe_nat,
+  int.nat_abs_mul_self, int.nat_abs_mul_self, complex.norm_sq]; simp
+
+@[simp] lemma nat_cast_complex_norm (x : ℤ[i]) : (x.norm : ℂ) = (x : ℂ).norm_sq :=
+by rw [← of_real_nat_cast, nat_cast_real_norm]
+
+@[simp] lemma norm_mul (x y : ℤ[i]) : norm (x * y) = norm x * norm y :=
```

Comment:
These statements are true in `zsqrtd` for general `d`.

Reply:
I agree that this is a good idea. Having the general notion of norm for `zsqrtd` opens the door to lots of elementary results regarding binary quadratic forms similar to the one you proved. Do you mind making that generalization?

Reply:
I've made the generalization. I think it would also be good to move `zsqrtd` to `data`, but that should probably be done separately so the git diff is clearer.